### PR TITLE
textDocument/completion support via gocode

### LINKED ISSUE #27: Feature request: Completion provider
https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#completion-request

### LINKED ISSUE #77: Support textDocument/completion

### LINKED ISSUE #201: Code completion.
Why is auto complete / code completion not supported? If it is why is it not shown in the graph on langserver.org ?

### LINKED ISSUE #215: CompletionItem.TextEdit should be a pointer
The current definition is:

```
type CompletionItem struct {
...
	TextEdit      TextEdit   `json:"textEdit,omitempty"`
...
}
```

It should be:

```
type CompletionItem struct {
...
	TextEdit      *TextEdit   `json:"textEdit,omitempty"`
...
}
```

Otherwise you *have* to provide an edit. The default `TextEdit` causes VSCode to log errors and doesn't show the completions.

Great package by the way, thanks!

### PR DESCRIPTION
**Pull request by Contextualist**
Support `textDocument/completion` by vendoring nsf/gocode. Gocode maintains a separated package cache, thus this implementation is not optimal for memory.

Known issues:
- No VFS support, because gocode uses exclusively .a binary cache
- ~Vendor import gets ignored, probably a bug of gocode (nsf/gocode#485)~
- ~Built-in types get ignored~

Commit ff980ce also fixes #215 .
This closes #27, closes #77, and closes #201

### CODE REVIEW

**Comment by keegancsmith** [comment]
Thanks!! I'll take a look at this tomorrow. cc @sqs @slimsag

**Comment by keegancsmith** [comment]
Apologies I won't have time today, I've put it onto my TODO list to review on Monday though.

**Comment by Contextualist** [comment]
@keegancsmith That's fine. I also need some time to test the feature with some corner cases as well as with real-world setting.

File: langserver/handler.go [hunk 1]
```diff
@@ -72,6 +73,13 @@ type LangHandler struct {
 
 // reset clears all internal state in h.
 func (h *LangHandler) reset(init *InitializeParams) error {
+	for _, k := range init.Capabilities.TextDocument.Completion.CompletionItemKind.ValueSet {
+		if k == lsp.CIKConstant {
+			CIKConstantSupported = lsp.CIKConstant
```
Comment (keegancsmith) [inline langserver/handler.go]: you are setting a global variable here, but langserver supports multiple instances running at once in the same process.

**Comment by keegancsmith** [comment]
@ramya-rao-a you may be interested in trying out the completion support in go-langserver this PR added. You need to ensure you are on latest master and run the langserver with the flag `-gocodecompletion`.

Thanks again @Contextualist!

**Comment by ramya-rao-a** [comment]
nice, will give it a try

**Comment by ramya-rao-a** [comment]
This works pretty well, nice job @Contextualist 

There are a few extra things that the Go extension to VS Code does for completions apart from what `gocode` returns. Once I figure out how they can play well with the basic completions from the language server, I can include this.
